libs/src/services/solanaWeb3Manager/rewards.js

```diff
@@ -0,0 +1,767 @@
+    isOracle,
+    recipientEthAddress,
+    tokenAmount,
+    transferId,
+    oracleAddress
+  )
+
+  return Secp256k1Program.createInstructionWithEthAddress({
+    ethAddress: attestationMeta.ethAddress,
+    message: encodedSenderMessage,
+    signature: encodedSignature,
+    recoveryId,
+    instructionIndex,
+  })
+}
+
+// Misc
+
+/**
+ * Converts an eth address hex represenatation to an array of Uint8s in big endian notation
+ * @param {string} ethAddress
+ * @returns {Uint8Array}
+ */
+const ethAddressToArray = (ethAddress) => {
+  const strippedEthAddress = ethAddress.replace("0x", "")
+  return Uint8Array.of(...new BN(strippedEthAddress, "hex").toArray("be"))
+}
+
+/**
+ * Constructs a transfer ID
+ * @param {string} challengeId
+ * @param {string} specifier
+ * @returns {string}
+ */
+const constructTransferId = (challengeId, specifier) =>
+  `${challengeId}:${specifier}`
+
+/**
+ * Converts a BN to a Uint8Array of length 8, in little endian notation.
+ * Useful for when Rust wants a u64 (8 * 8) represented as a byte array.
+ *
+ * @param {BN} bn
+ */
+const padBNToUint8Array = (bn) => bn.toArray("le", 8)
+
+/**
+ * Constructs an attestation from inputs.
+ *
+ * @param {boolean} isBot
+ * @param {string} recipientEthAddress
+ * @param {BN} tokenAmount
+ * @param {string} transferId
+ * @param {string} oracleAddress
+ * @returns {Uint8Array}
+ */
+const constructAttestation = (
+  isBot,
+  recipientEthAddress,
+  tokenAmount,
+  transferId,
+  oracleAddress
+) => {
+  const userBytes = ethAddressToArray(recipientEthAddress)
+  const oracleBytes = ethAddressToArray(oracleAddress)
+  const transferIdBytes = encoder.encode(transferId)
+  const amountBytes = padBNToUint8Array(tokenAmount)
+  const items = isBot
+    ? [userBytes, amountBytes, transferIdBytes]
+    : [userBytes, amountBytes, transferIdBytes, oracleBytes]
+  const sep = encoder.encode("_")
+  const res = items.slice(1).reduce((prev, cur, i) => {
+    return Uint8Array.of(...prev, ...sep, ...cur)
+  }, Uint8Array.from(items[0]))
+  return res
+}
+
+/**
+ * Derives the Solana account associated with a given sender Eth address.
+ *
+ * @param {string} ethAddress
+ * @param {PublicKey} rewardManagerProgramId
+ * @param {PublicKey} rewardManagerAccount
+ * @returns {Promise<PublicKey>}
+ */
+const deriveSolanaSenderFromEthAddress = async (
+  ethAddress,
+  rewardManagerProgramId,
+  rewardManagerAccount
+) => {
+  const ethAddressArr = ethAddressToArray(ethAddress)
+  const encodedPrefix = encoder.encode(SENDER_SEED_PREFIX)
+
+  const [, derivedSender] = await findProgramAddressWithAuthority(
+    rewardManagerProgramId,
+    rewardManagerAccount,
+    new Uint8Array([...encodedPrefix, ...ethAddressArr])
+  )
+  return derivedSender
+}
+
+/**
+ * Puts an instruction in a serializable form that our relay can understand.
+ *
+ * @param {TransactionInstruction} instruction
+ */
+const prepareInstructionForRelay = (instruction) => ({
```

Comment:
maybe this should be a shared util? don't all the other instructions have to do a similar thing? we should have 1 canonical way of doing this

Reply:
yeah 100% should be shared, will moev